### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by art4ride**
a bunch of optimisations, to get rid of `Lists.insert_at(_,-1,_)` and regexes

### CODE REVIEW

File: lib/sql_dust.ex [hunk 1]
```diff
@@ -179,22 +179,24 @@ defmodule SqlDust do
...
-                              {sql, options} = prepend_path_aliases("(" <> sql <> ")", options)
-                              {List.insert_at(conditions, -1, [sql] |> Enum.concat(values)), options}
-                            end)
+    {conditions, options} =
+      conditions
+      |> Enum.reduce({[], options}, fn([sql | values], {conditions, options}) ->
+        {sql, options} = prepend_path_aliases("(" <> sql <> ")", options)
+        {[[sql | values] | conditions], options}
+      end)
+    conditions = :lists.reverse(conditions)
```
Comment (nulian) [inline lib/sql_dust.ex]: Probably the Enum.reverse better for keeping in the elixir style

**Comment by archan937** [comment]
There are several failing tests

File: lib/sql_dust/utils/path_utils.ex [hunk 2]
```diff
@@ -33,19 +34,19 @@ require IEx
   def prepend_path_aliases2(sql, options) when is_binary(sql) do
```
Comment (nulian) [inline lib/sql_dust/utils/path_utils.ex]: Can you rename these I know my fault with my modifications but was more proof of concept on how to get more performance.

File: lib/sql_dust/utils/path_utils.ex [hunk 3]
```diff
@@ -60,20 +61,15 @@ require IEx
     {excluded, aliases} = scan_excluded(sql)
```
Comment (nulian) [inline lib/sql_dust/utils/path_utils.ex]: This was the original function used by the select now still used by the joins maybe it's possible to make it redundent and reuse the other version.

**Comment by art4ride** [comment]
@archan937 this branch already had these failing tests
+ I added 2 more returning " AS " instead of " as ".
Shouldn't be a problem but maybe I'm missing something. Wanted to discuss shall I implement the old behaviour or keep things simple, since lowercase " as " is not a must for SQL query

File: lib/sql_dust/utils/path_utils.ex (hunk 2 above)
Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: Yes, sure, just wanted to keep it closer to your version to stay on the same page and discuss, before polishing it up.

File: lib/sql_dust.ex (hunk 1 above)
Comment (art4ride) [inline lib/sql_dust.ex]: Already discussed with @nulian . Will use it.

File: lib/sql_dust/utils/path_utils.ex [hunk 4]
```diff
@@ -147,66 +150,73 @@ require IEx
...
   defp derive_path_alias(path, options) do
-    case String.replace(path, "#{quotation_mark(options)}", "") do
+    case String.replace(path, get_quotation_mark(options), "") do
       "" -> String.downcase(String.at(options.resource.name, 0))
       _ -> path
     end
   end
 
-  def quotation_mark(%{adapter: :mysql}) do
+  def get_quotation_mark(%{adapter: :mysql}) do
```
Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: Credo was angry on me to have function name and variable name to be the same

Comment (archan937) [inline lib/sql_dust/utils/path_utils.ex]: Lol. Well better the variable name instead :)

File: lib/sql_dust/utils/path_utils.ex [hunk 5]
```diff
@@ -125,17 +128,17 @@ require IEx
...
     Enum.reduce(paths, {sql, options}, fn([path], {sql, options}) ->
       {path_alias, options} = prepend_path_alias(path, options, true)
-      {String.replace(sql, "{#{path}}", path_alias), options}
+      {String.replace(sql, "{" <> path <> "}", path_alias), options}
     end)
   end
 
   def prepend_path_alias(path, options, cascade \\ false) do
-    {path, column} = dissect_path(path, options)
+    {path, column} = do_dissect_path(path, options)
```
Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: I split it up into 2 functions, to not perform list |>  join(".")  |> split(".") as it was previously

File: lib/sql_dust/utils/path_utils.ex (hunk 4 above)
Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: I thought of that first. It was a lot more changes to add (:

Comment (archan937) [inline lib/sql_dust/utils/path_utils.ex]: Hmmm. That's not really a valid reason to rename a function. Hahaha :)

Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: Ok, you got me :)

File: lib/sql_dust/utils/path_utils.ex (hunk 3 above)
Comment (art4ride) [inline lib/sql_dust/utils/path_utils.ex]: The function was renamed. Comment referenced to a different line number

File: lib/sql_dust/utils/scan_utils.ex [hunk 6]
```diff
@@ -50,13 +50,11 @@ defmodule SqlDust.ScanUtils do
...
-        match = hd(match)
-        case match do
-          "" -> quoted
-          _ -> List.insert_at(quoted, -1, match)
-        end
+      |> Enum.reduce([], fn
+        ([""|_], quoted) -> quoted
+        ([match|_], quoted) -> [match | quoted]
       end)
+      |> :lists.reverse
```
Comment (nulian) [inline lib/sql_dust/utils/scan_utils.ex]: Another reverse to change :)

File: mix.exs [hunk 7]
```diff
@@ -13,11 +13,13 @@ defmodule SqlDust.Mixfile do
   end
 
   def application do
-    [applications: [:logger]]
+    [applications: [:logger, :benchfella]]
   end
 
   defp deps do
     [
+      {:exprof, "~> 0.2.0"},
```
Comment (nulian) [inline mix.exs]: If we keep them in they should be optional so people aren't forced to add them to the project if possible.

Comment (art4ride) [inline mix.exs]: I think we need to set `only: [:dev, :test]` options for these ones

**Pull request merged**